cekit/descriptor/image.py

```diff
@@ -45,48 +50,318 @@ def __init__(self, descriptor, artifact_dir):
+        self._descriptor['modules'] = Modules(self._descriptor.get('modules', {}), self.path)
+        self._descriptor['packages'] = Packages(self._descriptor.get('packages', {}))
         if 'osbs' in self._descriptor:
             self._descriptor['osbs'] = Osbs(self._descriptor['osbs'])
         self._descriptor['volumes'] = [Volume(x) for x in self._descriptor.get('volumes', [])]
+
+        # make sure image declarations override any module definitions
+        self._image_overrides = { 'artifacts': self._to_dict(self.artifacts), 'modules': self._to_dict(self.modules.install) }
+        self._all_artifacts = self._to_dict(self.artifacts)
+
+    def process_defaults(self):
+        """Prepares default values before rendering"""
+        if not self.run:
+            self.run = Run({})
+
+        # do we want to force a user?
+        if 'user' not in self.run:
+            self.run._descriptor['user'] = cekit.DEFAULT_USER
+
+    @property
+    def name(self):
+        return self.get('name')
+
+    @name.setter
+    def name(self, value):
+        self._descriptor['name'] = value
+
+    @property
+    def version(self):
+        return self.get('version')
+
+    @version.setter
+    def version(self, value):
+        self._descriptor['version'] = value
+
+    @property
+    def release(self):
+        return self.get('release')
+
+    @release.setter
+    def release(self, value):
+        self._descriptor['release'] = value
+
+    @property
+    def base(self):
+        return self.get('from')
+
+    @base.setter
+    def base(self, value):
+        self._descriptor['from'] = value
+
+    @property
+    def description(self):
+        return self.get('description')
+
+    @description.setter
+    def description(self, value):
+        self._descriptor['description'] = value
+
+    @property
+    def labels(self):
+        return self.get('labels', [])
+
+    @property
+    def envs(self):
+        return self.get('envs', [])
+
+    @property
+    def ports(self):
+        return self.get('ports', [])
+
+    @property
+    def run(self):
+        return self.get('run')
+
+    @run.setter
+    def run(self, value):
+        self._descriptor['run'] = value
+
+    @property
+    def all_artifacts(self):
+        return self._all_artifacts.values()
+
+    @property
+    def artifacts(self):
+        return self.get('artifacts', [])
+
+    @property
+    def modules(self):
+        return self.get('modules', Modules({}, self._artifact_dir))
+
+    @property
+    def packages(self):
+        return self.get('packages', Packages({}))
+
+    @property
+    def osbs(self):
+        return self.get('osbs', Osbs({}))
+
+    @osbs.setter
+    def osbs(self, value):
+        self._descriptor['osbs'] = value
+
+    @property
+    def volumes(self):
+        return self.get('volumes', [])
+
+    @property
+    def help(self):
+        return self.get('help', {})
+
+    def apply_image_overrides(self, overrides):
```

Comment:
why we have different logic for applying overrides? it seems like code duplication. Its even using merge() method from descriptor.

Reply:
This is because the logic is different.  `apply_image_overrides()` applies the overrides specified on the command line, while `apply_module_overrides()` applies overrides coming from the modules.

If you look at the logic for `apply_module_overrides()`, you will see that it overrides the artifact definitions and module versions in the modules, based on what was specified in the image and/or overrides.  It also collects any package repositories specified on the modules (which should be considered deprecated, as we should only be declaring repositories on the image/overrides).  Lastly, it collects `run` settings, which will be merged into the image after all the modules have been processed.

There are a couple of reasons for processing things in this fashion.

1. It allows module versions to be overridden without affecting the order in which the modules are applied.
2. It keeps the module and image objects intact.  The only changes are the overrides that are applied, e.g. module version changes and artifact definitions, and in the case of the image, name, from, labels, envs, and the union of all run declarations.
3. Notice the order of the merge operations is reversed.

HTH

Reply:
Ok, but can it be move to Overrides.merge() Module.merge() respectively so its collocated with the descriptor?

Reply:
It's not merging, so no.  That's the purpose of this change.  There is no need to merge modules into the image.  The only merging that is being done is overrides into image, artifacts (from image and overrides) into modules, and the run field.

This requires a bit of a change of thinking.  With this, the overrides only override what is already specified, e.g. it can no longer be used to add modules or artifacts to an image, but can override what is already specified.  The other fields in the overrides (e.g. labels, etc.) only apply to the image itself.  The only thing modules override in the image is the run field, as there can be only one run field for the image.  This allows modules to configure run.  Other than that, everything is self contained.